cmd/zfs_object_agent/zettaobject/src/pool_destroy.rs

```
@@ -0,0 +1,584 @@
+        if buffer.is_empty() {
+            let this: Self = DestroyingCachePhys {
+                pools: HashMap::new(),
+            };
+            return Ok(this);
+        }
+        let destroying_cache: DestroyingCachePhys = serde_json::from_str(&buffer)?;
+
+        Ok(destroying_cache)
+    }
+
+    /*
+     * Write out a list of DestroyingCacheItemPhys as json so that if the agent restarts,
+     * pools that were in the process of being destroyed can be continued. To ensure
+     * that an agent crash does not leave the cache file partially written, we always
+     * write to a temp file and rename it atomically to replace the cache file.
+     * If an error is encountered while creating the file, we log an error and return.
+     */
+    async fn try_write(&self) {
+        trace!("Writing out destroy cache file.");
+
+        match Self::open_cache_file().await {
+            Ok((_, filename)) => {
+                let temp_filename = format!("{}.{}", filename, process::id());
+
+                match fs::write(
+                    &temp_filename,
+                    serde_json::to_string_pretty(&self).unwrap().as_bytes(),
+                )
+                .await
+                {
+                    Ok(_) => {
+                        fs::rename(temp_filename, filename).await.unwrap();
+                    }
+                    Err(error) => {
+                        error!("Could not write to {}. {:?}", &temp_filename, error);
+                    }
+                }
+            }
+            Err(error) => {
+                error!("Error opening zpool destroy cache file {}.", error);
+            }
+        }
+    }
+}
+
+/*
+ * Mark a pool as destroyed by writing to its super object.
+ */
+async fn mark_pool_destroying(
+    object_access: &ObjectAccess,
+    guid: PoolGuid,
+    total_data_objects: u64,
+) {
+    // Check to make sure that the pool exists.
+    if !PoolPhys::exists(object_access, guid).await {
```

This function seems to be best effort (i.e. it can silently fail).  So I think we could handle this by returning if PoolPhys::get() fails, since we don't really care if it fails due to not being there vs having invalid contents.

---

Ok